backends/bmv2/jsonconverter.cpp

```diff
@@ -101,6 +103,90 @@ class ErrorCodesVisitor : public Inspector {
+    JsonConverter* converter;
+    using Input = std::vector<const IR::Expression *>;
+    std::map<const IR::Declaration_Instance *, Input> selector_input_map{};
+
+ public:
+    explicit SharedActionSelectorCheck(JsonConverter* converter) : converter(converter)
+    { CHECK_NULL(converter); }
+
+    const Input &get_selector_input(const IR::Declaration_Instance* selector) const {
+      return selector_input_map.at(selector);
+    }
+
+    bool preorder(const IR::P4Table* table) override {
+        auto v1model = converter->v1model;
+        auto refMap = converter->refMap;
+        auto typeMap = converter->typeMap;
+
+        auto implementation = table->properties->getProperty(
+            v1model.tableAttributes.tableImplementation.name);
+        if (implementation == nullptr) return false;
+        if (!implementation->value->is<IR::ExpressionValue>()) {
+          ::error("%1%: expected expression for property", implementation);
+          return false;
+        }
+        auto propv = implementation->value->to<IR::ExpressionValue>();
+        if (!propv->expression->is<IR::PathExpression>()) return false;
+        auto pathe = propv->expression->to<IR::PathExpression>();
+        auto decl = refMap->getDeclaration(pathe->path, true);
+        if (!decl->is<IR::Declaration_Instance>()) {
+            ::error("%1%: expected a reference to an instance", pathe);
+            return false;
+        }
+        const auto &apname = decl->externalName();
+        auto dcltype = typeMap->getType(pathe, true);
+        if (!dcltype->is<IR::Type_Extern>()) {
+            ::error("%1%: unexpected type for implementation", dcltype);
+            return false;
+        }
+        auto type_extern_name = dcltype->to<IR::Type_Extern>()->name;
+        if (type_extern_name != v1model.action_selector.name) return false;
+
+        auto key = table->getKey();
+        Input input;
+        for (auto ke : *key->keyElements) {
+            auto mt = refMap->getDeclaration(ke->matchType->path, true)->to<IR::Declaration_ID>();
+            BUG_CHECK(mt != nullptr, "%1%: could not find declaration", ke->matchType);
+            if (mt->name.name != v1model.selectorMatchType.name) continue;
+            input.push_back(ke->expression);
+        }
+        auto decl_instance = decl->to<IR::Declaration_Instance>();
+        auto it = selector_input_map.find(decl_instance);
+        if (it == selector_input_map.end()) {
+            selector_input_map[decl_instance] = input;
+            return false;
+        }
+
+        // returns true if inputs are the same, false otherwise
+        auto cmp_inputs = [](const Input &i1, const Input &i2) {
+            for (auto e1 : i1) {
+                auto cmp_e = [e1](const IR::Expression *e2) {
+                    // not the best solution but the best available one given that there is no "depp
+                    // comparison" of expressions for now.
+                    return e1->toString() == e2->toString();
```

Comment:
This looks dangerous. 
What is really the meaning of the action selector? Can you describe how it is supposed to work?
I would have done this transformation myself when I did the action profiles, but I didn't know exactly what the meaning of action selectors is.

The problem here is that the same expression may evaluate to a different result.
Consider this case:

```
bit<32> x;

table t1() { key = { a[x].f: exact; } ... }
table t2() { key = { a[x].f: exact; } ... }

apply {
    x = 0;
    t1.apply();
    x = 1;
    t2.apply();
}
```

Although both key expressions are identical, they evaluate to different results.
The converse problem may happen too, where different expressions evaluate to the same result, e.g., because they use different temporary names.